### DISCUSSION TITLE
Fumbling Forward! Can anyone give me some guidance on how I read i2c from this device?

### DISCUSSION THREAD

**Discussion by wh33t** [micropython/micropython#10890] [ESP32]
So I have a [SparkX PHT](https://www.sparkfun.com/products/16298) - [SpecSheet](https://cdn.sparkfun.com/assets/b/a/f/b/f/MS8607_Datasheet.pdf) plugged into the qwiic port of my [Sparkfun ESP32 - Thing Plus C](https://www.sparkfun.com/products/20168)

Because I am using micropython none of their example scripts help me read the devices temperature or humidity, so I am attempting to do this step by step manually and treating it as a learning experience. 

```
# i2c_test.py
from machine import SoftI2C, Pin

i2c = SoftI2C(sda=Pin(21),scl=Pin(22)) # Sparkfun ESP32 uses SDA on Pin 21, SCL on Pin 22

devices = i2c.scan()

if len(devices) == 0:
    print("No i2c device !")
else:
    print('i2c devices found:',len(devices))
for device in devices:
    print("At address: ",hex(device))
```
Output

> i2c devices found: 3
> At address:  0x36
> At address:  0x40
> At address:  0x76

Now from what I understand so far, I want to read a value from device 0x76, but I when I try 

`
i2c.readfrom(0x76,0xA0)
`

I get this output

> Traceback (most recent call last):
>   File "<stdin>", line 15, in <module>
> OSError: [Errno 19] ENODEV

And it's clear to me I'm missing some fundamental knowledge here. 

Would you mind helping me fill in the gaps here? Not just what code to write, but how to gathering the required information from the specsheet for the PHT sensor?

Please and thank you so much!

**Comment by peterhinch** [comment]
You need a device driver for this chip. If you look at [Sparkfun's Arduino library](https://github.com/sparkfun/SparkFun_PHT_MS8607_Arduino_Library/blob/main/src/SparkFun_PHT_MS8607_Arduino_Library.cpp) you'll see it comprises a substantial C++ program. This needs to be translated to MicroPython. Unless, of course, someone has already done this.

**Reply by scruss** [reply]
There's a sort-of MicroPython driver for the sensor, but it looks like it would need unpacking from some [Digi XBee 3](https://www.digi.com/products/embedded-systems/digi-xbee/rf-modules/2-4-ghz-rf-modules/xbee3-zigbee-3#specifications) hardware specifics: [Xbee3-MicroPython/TE_MS8607_i2c.py at master · eewiki/Xbee3-MicroPython](https://github.com/eewiki/Xbee3-MicroPython/blob/master/samples/TE_MS8607_i2c.py)

There's also a CircuitPython driver that might make a start for porting: [Adafruit_CircuitPython_MS8607/adafruit_ms8607.py at main · adafruit/Adafruit_CircuitPython_MS8607](https://github.com/adafruit/Adafruit_CircuitPython_MS8607/blob/main/adafruit_ms8607.py)

Unlike some Qwiic boards, this one seems to expose the I2C sensor at the expected addresses: this is good. I like the idea of Qwiic --- I2C devices that can identify themselves to the user through a common ID and control address --- but found SparkFun's expectations a little too cumbersome to work with under MicroPython

**Comment by wh33t** [comment]
Appreciate the help guys. 

I just watched a great [little video](https://www.youtube.com/watch?v=SJTk7V7iC1I) on i2c and temperature sensors. 

It looks as though getting the temperature and humidity should be a matter of reading and writing bits of data from and to registers on the device using `readfrom_mem` and `writeto_mem` functions, then doing some conversion on the values.

My problem is that I don't understand how to read the spec sheet properly for this sensor, I have many questions and I am guessing this forum isn't the place to fill in those gaps? Could you recommend any terms, guides or videos to watch on how to understand that kind of spec sheet?

**Reply by Fisherman48** [reply]
Yes there is enought info in the datasheet to get temp/hum/pres but a lot of work to implement the calculations to get result within to specs. Easyer to port the python code ex from scruss. 

An faster approach to read temp/hum/pres, and not take the challange to learn a lot,  is to change device and use BMP/E 280 with the driver and tuturial from :
https://randomnerdtutorials.com/micropython-sensor-readings-email-esp32-esp826/

Br Stefan

**Reply by wh33t** [reply]
> Yes there is enought info in the datasheet to get temp/hum/pres but a lot of work to implement the calculations to get result within to specs. Easyer to port the python code ex from scruss.
> 
> An faster approach to read temp/hum/pres, and not take the challange to learn a lot, is to change device and use BMP/E 280 with the driver and tuturial from : https://randomnerdtutorials.com/micropython-sensor-readings-email-esp32-esp826/
> 
> Br Stefan

Thanks, I had considered switching to a sensor that already has a micropython library written for it. I noticed that [DHT20](https://www.sparkfun.com/products/18364) is cheap, and I believe [this](https://github.com/flrrth/pico-dht20) python library will work for it. The downside is that I believe I would have to solder it to the esp32 board and I have never done such a thing, it's why I chose the qwiic system to begin with. I'm torn between wanting to gain the knowledge to be good at this sort of thing or just buying as many off the shelf components so I can throw devices together quickly to immediately solve problems. 

While I've got eyes here on this thread, I'm curious, is it common for people who work with embedded development to be **able** to read datasheets and write their own drivers for things? I presume it's not that common for people to actually write their own drivers, but I'm curious how many people are capable of it. I've worked in tech and wrote many hundreds of thousands of lines of PHP, SQL, BASH and other scripting languages for almost two decades and never had to actually work with binary or hex in any meaningful way.

I think I may hire a local wizard to hold my hand and walk me through the creation of a very simple script that can read the sensor data according to that data sheet. I am intimidated by all of the binary/hex and bit shifting and I don't think I'll be able to learn that stuff online.

**Reply by Carglglz** [reply]
@wh33t This thread from the forum may help [Writing A Sensor Library/Driver](https://forum.micropython.org/viewtopic.php?f=15&t=9438)

**Reply by peterhinch** [reply]
> I'm curious, is it common for people who work with embedded development to be able to read datasheets and write their own drivers for things?

I would expect a professional firmware developer to be able to do this. However one of the aims of MicroPython is to bring firmware development to a wider audience. People who can write Python but who might lack the level of experience to write device drivers. That is why MicroPython includes some official drivers and why some users publish drivers for others to use.

Knowledge of binary/hex can help in writing efficient code: it's always worth learning.

**Reply by Fisherman48** [reply]
> I think I may hire a local wizard to hold my hand and walk me through the creation of a very simple script that can read the sensor data according to that data sheet. I am intimidated by all of the binary/hex and bit shifting and I don't think I'll be able to learn that stuff online.

Comming from opposite world 'I've worked with hw development/measurement and soldered thousands of connections ' and assembler bit oriented world -  my challange is to grasp all object thinking with class, self ,  declaration of vars lot of text ... Now learning by reading modules/ drivers.

I like your ambition and I think, why start to learn to read and understand both i2c protocoll and hw register contents/specs of this very complex module?
Start with a less complicated like a DS1307 real time clock or LCD display with just a few registers.

Br /Stefan
